### PR TITLE
Added conversion functions between Geometry and GeoJSON format. 

### PR DESCRIPTION
**Pull request by walterddr**
Test plan - Unit tests added in `TestGeoFunctions` for forward and backward conversion between JSON and Geometry format.

```
== RELEASE NOTES ==

General Changes
* Added :func:`geometry_from_geojson` and :func:`geometry_as_geojson` to convert geometries from and to GeoJSON format.
```

### CODE REVIEW

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 1]
```diff
@@ -87,17 +87,7 @@
...
-import static com.facebook.presto.geospatial.GeometryUtils.createJtsEmptyPolygon;
-import static com.facebook.presto.geospatial.GeometryUtils.createJtsLineString;
-import static com.facebook.presto.geospatial.GeometryUtils.createJtsMultiPoint;
-import static com.facebook.presto.geospatial.GeometryUtils.createJtsPoint;
-import static com.facebook.presto.geospatial.GeometryUtils.flattenCollection;
-import static com.facebook.presto.geospatial.GeometryUtils.getGeometryInvalidReason;
-import static com.facebook.presto.geospatial.GeometryUtils.getPointCount;
-import static com.facebook.presto.geospatial.GeometryUtils.jtsGeometryFromWkt;
-import static com.facebook.presto.geospatial.GeometryUtils.wktFromJtsGeometry;
+import static com.facebook.presto.geospatial.GeometryUtils.*;
```
Comment (jagill) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: In general, importing `*` is discouraged.  Make sure the imports are expanded.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 2]
```diff
@@ -231,6 +233,21 @@ public static boolean isPointOrRectangle(OGCGeometry ogcGeometry, Envelope envel
         return true;
     }
 
+    public static org.locationtech.jts.geom.Geometry jtsGeometryFromJson(String json)
+    {
+        try {
+            return new GeoJsonReader().read(json);
+        }
+        catch (ParseException | IllegalArgumentException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid WKT: " + e.getMessage(), e);
```
Comment (jagill) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java:242]: It's not a WKT, so the message should reference GeoJSON.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 3]
```diff
@@ -1266,4 +1267,26 @@ private void assertGeomFromBinary(String wkt)
     {
         assertFunction(format("ST_AsText(ST_GeomFromBinary(ST_AsBinary(ST_GeometryFromText('%s'))))", wkt), VARCHAR, wkt);
     }
+
+    @Test
+    public void testGeometryJsonConversion() throws Exception {
```
Comment (jagill) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Please add tests for:
1. Empty Geometries (of all types)
2. Deserializing valid GeoJSON strings
3. Deserializing invalid GeoJSON strings

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 4]
```diff
@@ -1266,4 +1267,26 @@ private void assertGeomFromBinary(String wkt)
...
+        assertGeoFromJson("LINESTRING (0 0, 1 2, 3 4)");
+        assertGeoFromJson("MULTILINESTRING ((1 1, 5 1), (2 4, 4 4))");
+        assertGeoFromJson("POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))");
+        assertGeoFromJson("POLYGON ((0 0, 0 3, 3 3, 3 0, 0 0), (1 1, 2 1, 2 2, 1 2, 1 1))");
+        assertGeoFromJson("MULTIPOLYGON (((1 1, 1 3, 3 3, 3 1, 1 1)), ((2 4, 2 6, 6 6, 6 4, 2 4)))");
+        assertGeoFromJson("GEOMETRYCOLLECTION (POINT (1 2), LINESTRING (0 0, 1 2, 3 4), POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0)))");
+
+        // invalid geometries
+        assertGeoFromJson("MULTIPOINT ((0 0), (0 1), (1 1), (0 1))");
+        assertGeoFromJson("LINESTRING (0 0, 0 1, 0 1, 1 1, 1 0, 0 0)");
```
Comment (jagill) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java:1285]: Also include a self-crossing LineString, like `(0 0, 1 1, 1 0, 0 1)`.

**Comment by jagill** [comment]
Also, make sure you work out the CLA/commit having the author issue.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java (hunk 3 above)
Comment (jagill) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: 3. Should include valid JSON but invalid GeoJSON, as well as invalid JSON.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 5]
```diff
@@ -1266,4 +1266,62 @@ private void assertGeomFromBinary(String wkt)
...
+    @Test
+    public void testGeometryJsonConversion()
+    {
+        // empty geometries
+        // JTS library has trouble creating correct JSON for empty, non-multi geometry,
+        // see https://github.com/locationtech/jts/issues/411
+        // the following tests has been commented out.
+//        assertGeoFromJson("POINT EMPTY");
+//        assertGeoFromJson("LINESTRING EMPTY");
+//        assertGeoFromJson("POLYGON EMPTY");
```
Comment (walterddr) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java:1279]: This is going to be a problem. I will see if there's any quick fix or workarounds we can do.

Comment (jagill) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java:1279]: It looks like the advise is that empty atomic (non-multi) geometries should return `null`: http://lists.geojson.org/pipermail/geojson-geojson.org/2013-October/000839.html

We can implement this by checking for those cases explicitly, and having `GeometryUtils.jsonFromJtsGeometry` return an `Optional<String>`.  In GeoFunctions, if it's empty, return null.  Else return the contained value.

**Comment by jagill** [comment]
1. I added a suggestion about handling atomic empty geometries.
2. Please add descriptions of the new functions in `presto-docs/.../geospatial.rst`.
3. Change the release notes to be something like:

```
== RELEASE NOTES ==

General Changes
* Added :func:`geometry_from_geojson` and :func:`geometry_as_geojson` to convert geometries from and to GeoJSON format.
```

**Comment by rongrong** [comment]
@walterddr Can you change the commit message as @jagill suggested? You can probably change the title to `Add conversion functions between Geometry and GeoJSON format` and put the two function names in commit message. Thanks!

**Pull request merged**